### PR TITLE
Move 2D and 3D resources to their own folders

### PR DESCRIPTION
**Pull request by aaronfranke**
Similarly to how we have the `scene/2d` folder and `scene/3d` folder for 2D and 3D nodes, this PR moves 2D and 3D resources to `scene/resources/2d` and `scene/resources/3d`. Also, I moved the skeleton modification resources to another subfolder, `scene/resources/2d/skeleton` since there were 18 2D skeleton modification resource files. Resources that are shared between 2D and 3D are still in `scene/resources`.

* This gives us the ability to easily exclude the resources in these folders when disabling 2D or 3D (but as of this PR, they are always included).

* In terms of compiled size, as of the current master, the resources folder is the biggest one, so that's another reason that I think it would be nice to divide it up into 2D and 3D subfolders:

godot4-chart

### CODE REVIEW

**Comment by clayjohn** [comment]
SurfaceTool and MeshDataTool should be under /resources rather than /resources/3d. Both can be used in 2D as well as 3D. For example, they can be used with a MeshInstance2D or a MultiMeshInstance2D.

**Comment by KoBeWi** [comment]
Why is `cavas_item_material` not in 2D?

Also maybe we should also have folder for GUI resources (theme, stylebox etc.)

**Comment by aaronfranke** [comment]
@KoBeWi Anything needed by Control nodes (such as CanvasItemMaterial) doesn't go in the 2D folder.

Making a folder for GUI resources may also be a good idea. I'm not sure if I want this PR's scope to grow too much though.

**Review by KoBeWi — APPROVED** [review]
I think the file moves in the PR make sense. We could make a couple of follow-ups to improve organization further (e.g. GUI folder), but it's a good start.

The downside is of course that it may break many PRs and rebasing after file renames is nasty from my experience.

**Comment by Zireael07** [comment]
Hopefully this gets merged before 4.0... been rebased tons of times already

**Comment by fire** [comment]
@aaronfranke do you remember how we generated those photos?

Back to topic: I remember this was not a minor change to merge because it causes all the godot engine prs to break.

**Comment by aaronfranke** [comment]
@fire https://docs.google.com/spreadsheets/d/1aWmWiA6MWoHr4Mgg22tybDFPa6BJX5oHTSjfN21plyY

**Comment by Zireael07** [comment]
Seriously, it's accumulating stuff and sitting there all this time and still not merged...

**Comment by Thed13** [comment]
Any update on this ?

**Comment by aaronfranke** [comment]
Awhile ago this was brought up in a meeting (should've written this down then, but I forgot), and reduz wondered if instead we should have `2d/resources` instead of `resources/2d`. The challenge here is that there are some resources that are neither 2D or 3D, so we would still need a `resources` folder. I am proposing this PR as it is because this is the path of least resistance vs the existing codebase, since these files are already in `resources`, this PR just adds another subfolder, but I am good with either approach, so whatever is preferred we can do that.

**Review by akien-mga — APPROVED** [review]
After discussion with maintainers, I agree with the change.

Conceptually, this is fine, a 2D/3D split marginally helps make the code organization better, but also paves the way to more possibility to decrease build size in custom 2D/3D only builds.

My main concern is with breaking compatibility on the include paths for C++ modules which use those Godot resources.
- For modules which only track a single Godot version (like in-house C++ modules for game-specific logic, etc.), it's fine, the devs can just do the same refactoring when updating to Godot 4.3.
- For general-purpose open source modules which aim to support multiple Godot branches, it's a bit trickier. They can still wrap things with `#if VERSION_HEX >= 0x040300` checks, though that implies including `core/version.h` in all affected files. Thankfully with GDExtension/godot-cpp, such modules are less common than they were in 3.x, so I think the tradeoff is acceptable (compatibility is not broken for godot-cpp as their the include paths are all flattened to a single folder).

File: scene/register_scene_types.cpp [hunk 1]
```diff
@@ -148,19 +167,16 @@
...
-#include "scene/resources/concave_polygon_shape_2d.h"
-#include "scene/resources/convex_polygon_shape_2d.h"
 #include "scene/resources/curve_texture.h"
 #include "scene/resources/environment.h"
 #include "scene/resources/font.h"
 #include "scene/resources/gradient.h"
 #include "scene/resources/gradient_texture.h"
 #include "scene/resources/image_texture.h"
 #include "scene/resources/immediate_mesh.h"
+#include "scene/resources/importer_mesh.h"
```
Comment (AThousandShips) [inline scene/register_scene_types.cpp:179]:
```suggestion
```
Was moved below, is 3D only

File: scene/register_scene_types.cpp [hunk 2]
```diff
@@ -265,20 +265,22 @@
...
+#include "scene/resources/3d/sky_material.h"
+#include "scene/resources/3d/sphere_shape_3d.h"
+#include "scene/resources/3d/world_boundary_shape_3d.h"
+#include "scene/resources/environment.h"
 #include "scene/resources/importer_mesh.h"
-#include "scene/resources/mesh_library.h"
 #include "scene/resources/primitive_meshes.h"
-#include "scene/resources/separation_ray_shape_3d.h"
-#include "scene/resources/sphere_shape_3d.h"
 #include "scene/resources/world_3d.h"
```
Comment (AThousandShips) [inline scene/register_scene_types.cpp:283]: Should be moved to 3D resources I think

File: scene/register_scene_types.cpp [hunk 3]
```diff
@@ -265,20 +264,22 @@
...
+#include "scene/resources/3d/convex_polygon_shape_3d.h"
+#include "scene/resources/3d/cylinder_shape_3d.h"
+#include "scene/resources/3d/fog_material.h"
+#include "scene/resources/3d/height_map_shape_3d.h"
+#include "scene/resources/3d/mesh_library.h"
+#include "scene/resources/3d/separation_ray_shape_3d.h"
+#include "scene/resources/3d/sky_material.h"
+#include "scene/resources/3d/sphere_shape_3d.h"
+#include "scene/resources/3d/world_boundary_shape_3d.h"
+#include "scene/resources/environment.h"
```
Comment (AThousandShips) [inline scene/register_scene_types.cpp:279]:
```suggestion
```
Not 3D only, already included

File: scene/register_scene_types.cpp [hunk 4]
```diff
@@ -265,20 +265,22 @@
...
 #include "scene/resources/importer_mesh.h"
-#include "scene/resources/mesh_library.h"
 #include "scene/resources/primitive_meshes.h"
```
[the remark this answers is not in the document]
Comment (aaronfranke) [reply scene/register_scene_types.cpp:281]: 2D meshes exist, so I'm not sure.

[the remark this answers is not in the document]
Comment (AThousandShips) [reply scene/register_scene_types.cpp:281]: The primitive mesh type isn't exposed on non-3D builds, see below, it was moved down below for compilation error reasons

File: scene/register_scene_types.cpp (hunk 2 above)
Comment (aaronfranke) [reply scene/register_scene_types.cpp:282]: Compiling Godot without including World3D is quite tricky. I tried to only move things into the 3D subfolder when I am able to exclude it when compiling Godot without 3D.

Comment (AThousandShips) [reply scene/register_scene_types.cpp:282]: It's not exposed though

**Review by AThousandShips — APPROVED** [review]

**Pull request merged**